host-controller/src/main/java/org/jboss/as/host/controller/mgmt/HostControllerRegistrationHandler.java

```diff
@@ -346,9 +349,22 @@ public void execute(final OperationContext context, final ModelNode operation) t
                     registrationContext.failed(failure, SlaveRegistrationException.ErrorCode.INCOMPATIBLE_VERSION, failure.getMessage());
                     throw failure;
                 }
+                // Ensure feature stream of host is compatible
+                LocalHostControllerInfo domainInfo = domainController.getLocalHostInfo();
+                if (domainInfo.getFeatureStream() != hostInfo.getFeatureStream()) {
+                    OperationFailedException failure = HostControllerLogger.ROOT_LOGGER.incompatibleFeatureStream(domainInfo.getFeatureStream(), hostInfo.getFeatureStream());
+                    registrationContext.failed(failure, SlaveRegistrationException.ErrorCode.INCOMPATIBLE_VERSION, failure.getMessage());
+                    throw failure;
+                }
+                // Legacy hosts must use DEFAULT stream
+                if ((hostVersion.compareTo(ModelVersion.CURRENT) < 0) && (hostInfo.getFeatureStream() != FeatureStream.DEFAULT)) {
+                    OperationFailedException failure = HostControllerLogger.ROOT_LOGGER.incompatibleFeatureStreamForLegacyHost(hostInfo.getFeatureStream());
+                    registrationContext.failed(failure, SlaveRegistrationException.ErrorCode.INCOMPATIBLE_VERSION, failure.getMessage());
+                    throw failure;
+                }
```

Comment:
Optional: To save one comparison, you could place the if that checks the versions before and add an `else` to verify the other case. It could make both checks more independent, but this is a very minor detail and not strictly neccessary at all. I leave it under your criteria, you can resolve this conversation yourself.

Reply:
Hmm - I'm confused.  Which comparison is the redundant one?